tesseract/tesseract_planning/tesseract_motion_planners/test/ompl_constrained_planner_tests.cpp

```diff
@@ -0,0 +1,219 @@
+
+#include <ompl/geometric/planners/sbl/SBL.h>
+#include <ompl/geometric/planners/est/EST.h>
+#include <ompl/geometric/planners/kpiece/LBKPIECE1.h>
+#include <ompl/geometric/planners/kpiece/BKPIECE1.h>
+#include <ompl/geometric/planners/kpiece/KPIECE1.h>
+#include <ompl/geometric/planners/rrt/RRT.h>
+#include <ompl/geometric/planners/rrt/RRTConnect.h>
+#include <ompl/geometric/planners/rrt/RRTstar.h>
+#include <ompl/geometric/planners/rrt/TRRT.h>
+#include <ompl/geometric/planners/prm/PRM.h>
+#include <ompl/geometric/planners/prm/PRMstar.h>
+#include <ompl/geometric/planners/prm/LazyPRMstar.h>
+#include <ompl/geometric/planners/prm/SPARS.h>
+
+#include <ompl/util/RandomNumbers.h>
+
+#include <functional>
+#include <cmath>
+#include <gtest/gtest.h>
+TESSERACT_COMMON_IGNORE_WARNINGS_POP
+
+#include <tesseract/tesseract.h>
+#include <tesseract_environment/core/utils.h>
+#include <tesseract_motion_planners/ompl/ompl_motion_planner.h>
+#include <tesseract_motion_planners/ompl/ompl_planner_configurator.h>
+#include <tesseract_motion_planners/ompl/config/ompl_planner_constrained_config.h>
+#include <trajopt/problem_description.hpp>
+#include <trajopt_utils/logging.hpp>
+
+using namespace tesseract;
+using namespace tesseract_scene_graph;
+using namespace tesseract_collision;
+using namespace tesseract_environment;
+using namespace tesseract_geometry;
+using namespace tesseract_kinematics;
+using namespace tesseract_motion_planners;
+
+const static int SEED = 1;
+const static std::vector<double> start_state = { -0.5, 0.5, 0.0, -1.3348, 0.0, 1.4959, 0.0 };
+const static std::vector<double> end_state = { 0.5, 0.5, 0.0, -1.3348, 0.0, 1.4959, 0.0 };
+
+std::string locateResource(const std::string& url)
+{
+  std::string mod_url = url;
+  if (url.find("package://tesseract_support") == 0)
+  {
+    mod_url.erase(0, strlen("package://tesseract_support"));
+    size_t pos = mod_url.find('/');
+    if (pos == std::string::npos)
+    {
+      return std::string();
+    }
+
+    std::string package = mod_url.substr(0, pos);
+    mod_url.erase(0, pos);
+    std::string package_path = std::string(TESSERACT_SUPPORT_DIR);
+
+    if (package_path.empty())
+    {
+      return std::string();
+    }
+
+    mod_url = package_path + mod_url;
+  }
+
+  return mod_url;
+}
+
+class GlassUprightConstraint : public ompl::base::Constraint
+{
+public:
+  GlassUprightConstraint(const Eigen::Vector3d& normal, tesseract_kinematics::ForwardKinematics::Ptr fwd_kin)
+    : ompl::base::Constraint(fwd_kin->numJoints(), 1), normal_(normal), fwd_kin_(std::move(fwd_kin))
+  {
+  }
+
+  ~GlassUprightConstraint() override = default;
+
+  void function(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::Ref<Eigen::VectorXd> out) const override
+  {
+    Eigen::Isometry3d pose;
+    fwd_kin_->calcFwdKin(pose, x);
+
+    Eigen::Vector3d z_axis = pose.matrix().col(2).template head<3>().normalized();
+
+    out[0] = std::atan2(z_axis.cross(normal_).norm(), z_axis.dot(normal_));
```

Comment:
Can you clarify your method for calculating the angle error here? I don't see how the magnitude of the cross product is relevant in this case because geometrically  the area of the parallelogram between the two vectors. Also this would not be well-behaved in the case that the vectors are orthogonal (the dot product, i.e. the denominator, is zero)

Reply:
![github_pic](https://user-images.githubusercontent.com/9803128/75292185-5697b200-57e9-11ea-9465-3480779982dc.jpg)

Above is a picture of the proof. One assumption I made was the normal was of unit length and probable should check if the input normal vector is unit length and if not normalize it. 

> Also this would not be well-behaved in the case that the vectors are orthogonal (the dot product, i.e. the denominator, is zero)

This is the reason for using `atan2` which checks for of these issues and returns the absolute angle (Not the acute angle) betweep -pi and pi.

Reply:
Okay, I understand now. I personally would still suggest using `1.0 - z_axis.dot(normal_)` (or `z_axis.norm() * normal_.norm() - z_axis.dot(normal_)` if you don't want to assume unit vectors) to reduce complexity/understand-ability, but it looks to produce the same result

Reply:
I agree. This is for testing because it is easier to understand since the units are radians.